### ISSUE TITLE
Unnecessary_cast wrongly triggers on primitive methods of i32

### ISSUE THREAD

**Issue by SebastianRefsnes** [rust-lang/rust-clippy#11298]
### Summary

I tried to see if this bug existed on all the integer number types, and while I haven't exhaustively tested every primitive method, `count_ones()`, as I used in my example, is correctly handled on:
```
i8   , u8,
i16  , u16
 __  , u32
i64  , u64
i128 , u128
```

It seems to me that this only fails when `i` is i32

Other methods that fail with i32 are: (`count_zeros()`, `ilog2()`, `trailing_ones()`, `trailing_zeroes()`...) <- not exhaustive

### Lint Name

unnecessary_cast

### Reproducer

I tried this code:

```rust
fn main() {
    for i in 1..(1 << 4) {
        println!("{}", (i as i32).count_ones() << 2 );
    }
}

```

I saw this happen:

```
warning: casting to the same type is unnecessary (`i32` -> `i32`)
 --> src/main.rs:3:24
  |
3 |         println!("{}", (i as i32).count_ones() << 2 );
  |                        ^^^^^^^^^^ help: try: `i`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
  = note: `#[warn(clippy::unnecessary_cast)]` on by default
```

I expected to see this happen:

no warning, allowing clippy to go through with its suggestion for i32 would result in a compiler error

### Version

```text
rustc 1.71.1 (eb26296b5 2023-08-03)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: aarch64-apple-darwin
release: 1.71.1
LLVM version: 16.0.5
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by Centri3** [comment]
You would usually supply the type elsewhere: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9f995d9bb6e487b65e5ec5a0fcb7012b

It makes sense only `i32` would fail because `i` here is inferred to be `i32` due to default numeric fallback, supplying the suffix in the range like `1u32` then using `as u32` on it would result in the same error.

So, not really a FP as you can always supply the suffix (or type!) *somewhere*, but the suggestion shouldn't be auto-applied as it can break code.

**Comment by SebastianRefsnes** [comment]
> You would usually supply the type elsewhere: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9f995d9bb6e487b65e5ec5a0fcb7012b
> 
> It makes sense only `i32` would fail because `i` here is inferred to be `i32` due to default numeric fallback, supplying the suffix in the range like `1u32` then using `as u32` on it would result in the same error.
> 
> So, not really a FP as you can always supply the suffix (or type!) _somewhere_, but the suggestion shouldn't be auto-applied as it can break code.
> 
> The reason why the method call then fails is that these methods are shared between all integers, afaik any method call would cause this.

How is it not a false positive just because you can change some *other* part of your code to make it work?
The fact of the matter is that if you leave it as is, it compiles, but if you apply the suggested fix, it does not compile.

**Comment by Centri3** [comment]
It's not a FP because the general suggestion is correct; this is an unnecessary cast. There are other lints like this; applying the suggestion may break code, (or introduce a semantic change,) and thus shouldn't be applied automatically. But it is still an unnecessary cast, the lint's general advice is correct, you're just expected to apply it yourself and potentially fix the resulting errors.

Now, ideally, should the suggestion be to add the suffix instead of using an `as` cast? Yes, absolutely, but that would be tricky due to how many ways this can happen: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=88818b1f50a643ae3c8fbb8ad0fccf0d
